### PR TITLE
Added 'dps' in _default_settings in StrPrinter class to control decimal places of floats

### LINKED ISSUE #27033: Request for setting in the StrPrinter to control the decimal places of floats
The `StrPrinter` prints floats using mpmath's `mlib_to_str` and the decimal places `dps` is automatically selected based on the sympy float properties. We could allow the `dps` to be user controlled with something like:

```python
import sympy as sm
from mpmath.libmp import prec_to_dps, to_str as mlib_to_str


class MyPrinter(sm.StrPrinter):

    _default_settings = {
        "order": None,
        "full_prec": "auto",
        "sympy_integers": False,
        "abbrev": False,
        "perm_cyclic": True,
        "min": None,
        "max": None,
        "dps": None,
    }

    def _print_Float(self, expr):
        if self._settings['dps']:
            dps = self._settings['dps']
        else:
            dps = 0 if expr._prec < 5 else prec_to_dps(expr._prec)
        if self._settings["full_prec"] is True:
            strip = False
        elif self._settings["full_prec"] is False:
            strip = True
        elif self._settings["full_prec"] == "auto":
            strip = self._print_level > 1
        low = self._settings["min"] if "min" in self._settings else None
        high = self._settings["max"] if "max" in self._settings else None
        rv = mlib_to_str(expr._mpf_, dps, strip_zeros=strip, min_fixed=low, max_fixed=high)
        if rv.startswith('-.0'):
            rv = '-0.' + rv[3:]
        elif rv.startswith('.0'):
            rv = '0.' + rv[2:]
        if rv.startswith('+'):
            # e.g., +inf -> inf
            rv = rv[1:]
        return rv

f = sm.Function('f')
print(MyPrinter(settings={'dps': 3}).doprint(f(1.329294)))
```

```
f(1.33)
```

This would allow floats to print more compactly.

### PR DESCRIPTION
**Pull request by VectorNd**
#### References to other Issues or PRs
Fixes #27033

#### Brief description of what is fixed or changed
Added a {"dps" : None} pair in _default_settings in StrPrinter and accordingly made some change in the _print_Float method . 

#### Release Notes

* printing 
  * `dps` can now be passed as a keyword for StrPrinter to control the number of digits shown

### CODE REVIEW

**Comment by oscarbenjamin** [comment]
You don't need to close the pull request

**Comment by oscarbenjamin** [comment]
The code quality test has failed because of trailing whitespace:
https://github.com/sympy/sympy/actions/runs/10746113244/job/29806427330?pr=27043

The error message tells you which file and which line:

> FAILED sympy/testing/tests/test_code_quality.py::test_files - AssertionError: File contains trailing whitespace: /home/runner/work/sympy/sympy/sympy/printing/str.py, line 741.

You can delete the spaces at the end of the line or run the `bin/strip_whitespace` script. Once that is done you can push another commit to your branch which will update the pull request.

You do *not* need to close the pull request and open a new one.

File: sympy/printing/str.py [hunk 1]
```diff
@@ -736,10 +737,13 @@ def _print_mpq(self, expr):
 
     def _print_Float(self, expr):
         prec = expr._prec
-        if prec < 5:
-            dps = 0
+        if self._settings['dps']:
+            dps = self._settings['dps']
```
Comment (oscarbenjamin) [inline sympy/printing/str.py:741]: There is a space at the end of this line.

Comment (VectorNd) [reply sympy/printing/str.py:741]: okay thankyou

**Comment by oscarbenjamin** [comment]
The new failures are not caused by your changes here. They are because of an update to ruff.

I have a fix in gh-27040 but I am waiting for someone to merge it. Once that happens the failures seen here in this pull request will be resolved as well.

**Comment by VectorNd** [comment]
@oscarbenjamin  can you please tell me why is it failing now i have not made any change in any other file than str.py

**Comment by moorepants** [comment]
There are few things this will need:

1. unit tests
2. docstring explanation of the new setting'
3. figuring out what happens when this new setting is used alongside any of the other settings and documenting/testing for that

**Comment by VectorNd** [comment]
@moorepants  sir can you please merge it  ?

**Comment by moorepants** [comment]
> @moorepants sir can you please merge it ?

No, but I am willing to review it when I have time.

**Comment by VectorNd** [comment]
okay thankyou

File: sympy/printing/str.py [hunk 2]
```diff
@@ -1007,6 +1017,85 @@ def _print_Str(self, s):
         # Str does not to be printed same as str here
         return "%s(%s)" % (s.__class__.__name__, self._print(s.name))
 
+# class StrDecPrinter(StrPrinter):
```
Comment (asmeurer) [inline sympy/printing/str.py:1020]: Do not add commented out code like this. If the code is not ready yet, you can discuss it on the issue, but it shouldn't be added like this.

**Comment by VectorNd** [comment]
@asmeurer i have fixed the mistakes . can you please review it

File: sympy/printing/str.py [hunk 3]
```diff
@@ -736,10 +737,9 @@ def _print_mpq(self, expr):
 
     def _print_Float(self, expr):
         prec = expr._prec
-        if prec < 5:
-            dps = 0
-        else:
-            dps = prec_to_dps(expr._prec)
+        dps = self._settings.get('dps',None)
+        if dps is None:
+            dps = 0 if prec < 5 else prec_to_dps(expr._prec)
```
Comment (smichr) [inline sympy/printing/str.py:742]:
```suggestion
        dps = self._settings.get('dps',
              0 if prec < 5 else prec_to_dps(expr._prec)
```

**Comment by smichr** [comment]
It looks ok to me. I'll wait a while to see if anyone else has suggestions.

File: sympy/printing/str.py [hunk 4]
```diff
@@ -736,10 +737,8 @@ def _print_mpq(self, expr):
 
     def _print_Float(self, expr):
         prec = expr._prec
-        if prec < 5:
-            dps = 0
-        else:
-            dps = prec_to_dps(expr._prec)
+        dps = self._settings.get('dps',
+              0 if prec < 5 else prec_to_dps(expr._prec))
```
Comment (smichr) [inline sympy/printing/str.py:741]:
```suggestion
        dps = self._settings.get('dps', None)
        if dps is None:
            dps = 0 if prec < 5 else prec_to_dps(expr._prec))
```
```suggestion
        dps = self._settings.get('dps',
              0 if prec < 5 else prec_to_dps(expr._prec))
```

File: sympy/printing/str.py [hunk 5]
```diff
@@ -736,10 +737,9 @@ def _print_mpq(self, expr):
 
     def _print_Float(self, expr):
         prec = expr._prec
-        if prec < 5:
-            dps = 0
-        else:
-            dps = prec_to_dps(expr._prec)
+        dps = self._settings.get('dps', None)
+        if dps is None:
+            dps = 0 if prec < 5 else prec_to_dps(expr._prec))
```
Comment (smichr) [inline sympy/printing/str.py:742]:
```suggestion
            dps = 0 if prec < 5 else prec_to_dps(expr._prec)
```

**Comment by smichr** [comment]
If locally you could add a space between the comma and None and then squash all the commits into one and then force push here, that would be helpful. Locally, you have 9 commits -- double check -- so I would proceed like this in git:
```
git reset head~9
# make the edit
git commit -am 'add dps to StrPrinter'
git push -f VectorNd
```

**Comment by smichr** [comment]
Perfect! Thanks.

**Comment by moorepants** [comment]
Neither of these two points I raised above seem to have been resolved:

> 2. docstring explanation of the new setting'
> 3. figuring out what happens when this new setting is used alongside any of the other settings and documenting/testing for that


`str.py` has practically no documentation for the user to understand how to use any of the settings.

**Pull request merged**